pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+				glog.V(4).Infof("Setting reason for pending build to %q due to missing secret for %s", build.Status.Reason, buildDesc(build))
+			}
+		}
+	case kapi.PodRunning:
+		if build.Status.Phase != buildapi.BuildPhaseRunning {
+			update.setReason("")
+			update.setMessage("")
+			update.setPhase(buildapi.BuildPhaseRunning)
+			if pod.Status.StartTime != nil {
+				update.setStartTime(*pod.Status.StartTime)
+			}
+		}
+	case kapi.PodSucceeded:
+		if build.Status.Phase != buildapi.BuildPhaseComplete {
+			update.setReason("")
+			update.setMessage("")
+			update.setPhase(buildapi.BuildPhaseComplete)
+		}
+		if len(pod.Status.ContainerStatuses) == 0 {
+			// no containers in the pod means something went terribly wrong, so the build
+			// should be set to an error state
+			glog.V(2).Infof("Setting build %s to error state because its pod has no containers", buildDesc(build))
+			update.setPhase(buildapi.BuildPhaseError)
+			if build.Status.Reason == "" {
+				update.setReason(buildapi.StatusReasonBuildPodDeleted)
+				update.setMessage(buildapi.StatusMessageBuildPodDeleted)
+			}
+		} else {
+			for _, info := range pod.Status.ContainerStatuses {
+				if info.State.Terminated != nil && info.State.Terminated.ExitCode != 0 {
+					glog.V(2).Infof("Setting build %s to error state because a container in its pod has non-zero exit code", buildDesc(build))
+					update.setPhase(buildapi.BuildPhaseError)
+					break
+				}
+			}
+		}
+	case kapi.PodFailed:
+		if build.Status.Phase != buildapi.BuildPhaseFailed {
+			// If a DeletionTimestamp has been set, it means that the pod will
+			// soon be deleted. The build should be transitioned to the Error phase.
+			if pod.DeletionTimestamp != nil {
+				return podDeletedUpdate(), nil
+			}
+
+			update.setPhase(buildapi.BuildPhaseFailed)
+			update.setReason(buildapi.StatusReasonGenericBuildFailed)
+			update.setMessage(buildapi.StatusMessageGenericBuildFailed)
+		}
+	}
+	return update, nil
+}
+
+// handleCompletedBuild will only be called on builds that are already in a terminal phase however, their completion timestamp
+// has not been set.
+func (bc *BuildController) handleCompletedBuild(build *buildapi.Build, pod *kapi.Pod, podErr error) (*buildUpdate, error) {
+	// Make sure that the completion timestamp has not already been set
+	if build.Status.CompletionTimestamp != nil {
+		return nil, nil
+	}
+
+	update := &buildUpdate{}
+	var podStartTime *metav1.Time
+	if podErr != nil && pod != nil {
+		podStartTime = pod.Status.StartTime
+	}
+	setBuildCompletionTimestampAndDuration(build, podStartTime, update)
+
+	return update, nil
+}
+
+// updateBuild is the single place where any update to a build is done in the build controller.
+// It will check that the update is valid, peform any necessary processing such as calling HandleBuildCompletion,
+// and apply the buildUpdate object as a patch.
+func (bc *BuildController) updateBuild(build *buildapi.Build, update *buildUpdate, pod *kapi.Pod) {
```

Comment:
why does updateBuild not return error?  If the patch fails, shouldn't a requeue happen?